tests/test_directives.py

```diff
@@ -0,0 +1,270 @@
+        for dir in available_directives_list:
+            with pytest.raises(error_types[dir.name]):
+                dir.validation(invalid_values[dir.name])
+
+    def test_defaults_are_valid(self, available_directives_list):
+        for directive in available_directives_list:
+            directive.validation(directive.default)
+
+    def test_serial(self):
+        assert NP.serial(4, 2) == 4
+        assert NRANKS.serial(4, 2) == 4
+        assert NGPU.serial(4, 2) == 4
+        assert OMP_NUM_THREADS.serial(4, 2) == 4
+        assert EXECUTABLE.serial('Path1', 'Path2') == 'Path1'
+        assert WALLTIME.serial(4, 2) == 6
+        assert MEMORY.serial(4, 2) == 4
+        assert PROCESS_FRACTION.serial(0.4, 0.2) == 0.4
+
+    def test_parallel(self):
+        assert NP.parallel(4, 2) == 6
+        assert NRANKS.parallel(4, 2) == 6
+        assert NGPU.parallel(4, 2) == 6
+        assert OMP_NUM_THREADS.parallel(4, 2) == 6
+        assert EXECUTABLE.parallel('Path1', 'Path2') == 'Path1'
+        assert WALLTIME.parallel(4, 2) == 4
+        assert MEMORY.parallel(4, 2) == 6
+        assert PROCESS_FRACTION.parallel(0.4, 0.2) == 0.4
+
+    def test_finalize(self):
+        dict_directives = {'nranks': NRANKS.default, 'omp_num_threads': OMP_NUM_THREADS.default}
+        assert NP.finalize(2, dict_directives) == 2
+        dict_directives['nranks'] = 2
+        dict_directives['omp_num_threads'] = 4
+        assert NP.finalize(2, dict_directives) == 8
+        dict_directives['nranks'] = lambda x: x**2
+        assert NP.finalize(2, dict_directives) == 2
+
+    def test_manual_item_default(self, product_directive):
+        assert product_directive.default == 10
+
+    def test_manual_item_validation(self, product_directive):
+        try:
+            val = product_directive.validation(product_directive.default)
+            assert product_directive.default == val
+        except Exception:
+            raise ValueError("Invalid default value for product_directive")
+        assert product_directive.default == product_directive.validation(product_directive.default)
+        assert product_directive.validation(20) == 20
+        with pytest.raises(ValueError):
+            product_directive.validation(0)
+
+    def test_manual_item_serial(self, product_directive):
+        product_directive.serial(10, 20) == 10
+
+    def test_manual_item_parallel(self, product_directive):
+        product_directive.parallel(10, 20) == 10
+
+    def test_manual_item_finalize(self, product_directive):
+        asset_dict = {'free': False, 'discount': 5}
+        assert product_directive.finalize(50, asset_dict) == 45
+        asset_dict['free'] = True
+        assert product_directive.finalize(50, asset_dict) == 0
+
+
+class TestDirectives:
+    """
+    Tests for Directives Class
+    """
+    def test_get_directive(self, available_directives_list):
+        directives = Directives(available_directives_list)
```

Comment:
You could make this directives object a fixture, if you want since it is used in multiple places. I don't feel strongly either way.

Reply:
I too don't feel strongly either way. I think declaring the class inside a method is a good option here because of the "pass by reference" nature of the classes in python.